QModManager/API/SMLHelper/Assets/Spawnable.cs

```diff
@@ -0,0 +1,125 @@
+        protected Spawnable(string classId, string friendlyName, string description)
+            : base(classId, $"{classId}Prefab")
+        {
+            if (string.IsNullOrEmpty(classId))
+            {
+                Logger.Error($"ClassID for Spawnables must be a non-empty value.");
+                throw new Exception($"Error patching Spawnable");
+            }
+
+            this.FriendlyName = friendlyName;
+            this.Description = description;
+
+            CorePatchEvents += RegisterBasics;
+        }
+
+        /// <summary>
+        /// This event triggers <c>before</c> the core patching methods begins.
+        /// You can attach simple <seealso cref="PatchEvent"/> methods to this event if you want to run code <c>before</c> the any of the core patching methods begin.
+        /// </summary>
+        protected PatchEvent OnStartedPatching;
+
+        /// <summary>
+        /// The main patching methods are executed here.
+        /// This event should only be used by the SMLHelper QuickStart classes.
+        /// </summary>
+        internal PatchEvent CorePatchEvents;
+
+        /// <summary>
+        /// This event triggers <c>after</c> the core patching methods begins.
+        /// You can attach simple <seealso cref="PatchEvent"/> methods to this event if you want to run code <c>after</c> the core patching methods have finished.
+        /// </summary>
+        protected PatchEvent OnFinishedPatching;
+
+        /// <summary>
+        /// Starts all patching code in SMLHelper.
+        /// If <seealso cref="IsPatched"/> is <c>true</c> then this method is skipped to avoid duplicate patching.
+        /// </summary>
+        /// <seealso cref="OnStartedPatching"/>
+        /// <seealso cref="OnFinishedPatching"/>
+        public void Patch()
+        {
+            if (this.IsPatched)
+                return; // Already patched. Skip.
+
+            OnStartedPatching?.Invoke();
+
+            // Because invocation order isn't guaranteed by event handlers,
+            // we make sure the TechType is patched first before anything else that might require it.
+            this.TechType = TechTypeHandler.Singleton.AddTechType(ModName, this.ClassID, this.FriendlyName, this.Description, false);
+
+            CorePatchEvents.Invoke();
+
+            this.IsPatched = true;
+
+            OnFinishedPatching?.Invoke();
+        }
+
+        private void RegisterBasics()
+        {
+            PrefabHandler.RegisterPrefab(this);
+
+            string assetsFolder = this.AssetsFolder;
+            if (string.IsNullOrEmpty(assetsFolder))
+            {
+                Logger.Error($"AssetsFolder property for Spawnable instance of {this.ClassID} must have a non-empty value.");
+                throw new Exception($"Error patching Spawnable:{this.ClassID}");
+            }
+
+            SpriteHandler.RegisterSprite(this.TechType, $"./QMods/{assetsFolder.Trim('/')}/{this.IconFileName}");
```

Comment:
Oh god. Making paths with string literals is a no-no.

Reply:
I agree. Let's see if it's possible to have this be backwards compatible for all the mods currently using it which put all these assets into a separate folder that's at the same level as the DLL.